### ISSUE TITLE
make check on fedora with gcc-7 fails on ppc64, because of altivec

### ISSUE THREAD

**Issue by ingvarha** [jemalloc/jemalloc#1088]
When building 
gcc -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mno-altivec -c -D_GNU_SOURCE -D_REENTRANT -Iinclude -Iinclude -DJEMALLOC_UNIT_TEST -Itest/include -Itest/include -o test/src/btalloc.unit.o test/src/btalloc.c
In file included from test/include/test/jemalloc_test.h:33:0,
                 from test/src/btalloc.c:1:
/usr/lib/gcc/ppc64-redhat-linux/7/include/altivec.h:34:2: error: #error Use the "-maltivec" flag to enable PowerPC AltiVec support
 #error Use the "-maltivec" flag to enable PowerPC AltiVec support
  ^~~~~

Patching out explicit altivec usage  (HAVE_ALTIVEC) "fixes" the problem, and can be used as a workaround. I did this for jemalloc-3 too, I remember.

Compiling with -maltivec is probably not a real option here, as all ppc variants do not have altivec. The correct way is probably to use altivec only when it is available. Fedora will with by default build packages with gcc-7 on ppc64, with -mno-altivec.

See also issue #119

Compiling without altivec hides this bug too:

include/jemalloc/internal/../jemalloc.h:280:74: error: unknown type name '__bool'; did you mean '_Bool'?
 typedef void *(extent_alloc_t)(extent_hooks_t *, void *, size_t, size_t, bool *,
                                                                          ^
include/jemalloc/internal/../jemalloc.h:281:5: error: unknown type name '__bool'; did you mean '_Bool'?
     bool *, unsigned);
     ^
include/jemalloc/internal/../jemalloc.h:288:66: error: unknown type name '__bool'; did you mean '_Bool'?
 typedef bool (extent_dalloc_t)(extent_hooks_t *, void *, size_t, bool,
                                                                  ^
(...)

The following was snipped from a post from Jakub Jelinek on the fedora-devel list; the package in question here is compiled with g++, but hit the same problem

Ingvar

These days (for several GCC releases), powerpc as well as spu and recently
also s390 (z13) use conditional macros for bool, vector, pixel and _Bool,
which expand differently depending on what is the following token.
But the conditional macros are defined only in the non-strict C/C++ modes,
the above is compiled with -std=c++11 rather than -std=gnu++11, but then
one shouldn't include altivec.h if they want standard behavior.
Alternative is to include altivec.h and
#undef vector
#undef pixel
#undef bool
and then of course you need to use __vector, __pixel and __bool instead of
vector, pixel or bool if you want the Altivec-ish behavior rather than
standard.

**Comment by ingvarha** [comment]
This is jemalloc-5.0.1, by the way

**Comment by sharkcz** [comment]
It should be fixable with a real check for Altivec in configure, will try to come up with a patch

**Comment by sharkcz** [comment]
The Altivec handling in configure got removed in commit 1245faae os it's not a problem any more. Although it could be fixed with eg.
~~~~
diff -up jemalloc-5.0.1/configure.ac.altivec jemalloc-5.0.1/configure.ac
--- jemalloc-5.0.1/configure.ac.altivec 2017-12-14 15:23:20.423199027 +0100
+++ jemalloc-5.0.1/configure.ac 2017-12-14 15:17:14.145431853 +0100
@@ -399,7 +399,7 @@ case "${host_cpu}" in
        fi
        ;;
   powerpc*)
-       AC_DEFINE_UNQUOTED([HAVE_ALTIVEC], [ ])
+       AC_CHECK_HEADERS(["altivec.h"], AC_DEFINE_UNQUOTED([HAVE_ALTIVEC], [ ]), [])
        CPU_SPINWAIT='__asm__ volatile("or 31,31,31")'
        ;;
   *)
~~~~

**Comment by sharkcz** [comment]
A note for upstream - if you would like to add ppc64/ppc64le/s390x into your CI setup, let me know.

**Comment by davidtgoldblatt** [comment]
> if you would like to add ppc64/ppc64le/s390x into your CI setup, let me know.

Yeah, we would love to. Mind dropping by in https://github.com/jemalloc/jemalloc/issues/1044 ?

Closing this for now, since it seems like it's fixed by #1043. (Feel free to reopen if I missed something).
